src/main/java/faang/school/projectservice/service/StageService.java

```diff
@@ -0,0 +1,123 @@
+import faang.school.projectservice.model.TeamMember;
+import faang.school.projectservice.model.TeamRole;
+import faang.school.projectservice.model.stage.Stage;
+import faang.school.projectservice.model.stage.StageRoles;
+import faang.school.projectservice.repository.ProjectRepository;
+import faang.school.projectservice.repository.StageRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Component
+@RequiredArgsConstructor
+public class StageService {
+    private final StageRepository stageRepository;
+    private final ProjectRepository projectRepository;
+    private final StageMapper stageMapper;
+    public void makeStage(String stageName, Long projectId, List<StageRoles> stageRoles) {
+        List<Task> tasks = new ArrayList<>();
+        List<TeamMember> executors = new ArrayList<>();
+        makeStage(stageName, projectId, stageRoles, tasks, executors);
+    }
+
+    public void makeStage(String stageName, Long projectId, List<StageRoles> stageRoles, List<Task> tasks, List<TeamMember> executors) {
+        if (projectRepository.getProjectById(projectId).getStatus() == ProjectStatus.CANCELLED) {
+            throw new ProjectStatusException("You can not make stage for cancelled project!");
+        }
+        if (projectRepository.getProjectById(projectId).getStatus() == ProjectStatus.COMPLETED) {
+            throw new ProjectStatusException("You can not make stage for completed project!");
+        }
+        Stage stage = new Stage();
+        stage.setStageName(stageName);
+        stage.setProject(projectRepository.getProjectById(projectId));
+        stage.setStageRoles(stageRoles);
+        stage.setTasks(tasks);
+        stage.setExecutors(executors);
+        stageRepository.save(stage);
+    }
+
+    public List<StageDto> getStagesByStatus(Long projectId, TaskStatus status) {
+        List<Stage> projectStages = stageRepository.findAll().stream()
+                .filter(stage -> stage.getProject().getId() == projectId).toList();
+        return stageMapper.toDto(projectStages.stream()
+                .filter(stage -> {
+                    if (stage.getTasks().stream().filter(task -> task.getStatus().equals(status)).toList().isEmpty()) {
+                        return false;
+                    }
+                    return true;
+                })
+                        .toList());
+    }
+
+    public void deleteStage(Long stageId) {
+        Stage stage = stageRepository.getById(stageId);
+        List<Task> tasks = stage.getTasks();
+        for (Task task : tasks) {
+            task.setStatus(TaskStatus.CANCELLED);
+        }
+        stageRepository.delete(stage);
+    }
+
+    public void deleteStage(Long stageIdToDelete, Long stageIdToReceive) {
+        Stage stageToDelete = stageRepository.getById(stageIdToDelete);
+        Stage stageToReceive = stageRepository.getById(stageIdToReceive);
```

Comment:
why do you need stageToReceive ?

Reply:
It's a varaiant of deleting with transfering tasks to another stage

Reply:
But when I try to use stage.setTasks(List.of(task1, task2)) and then stage.getTasks(), there is an StackOverflowError
And there is no such error when try to setTasks with an empty list